### PR TITLE
Milestone 5

### PR DESCRIPTION
**Pull request by phylischumba**
### In this model we:
- Generated the friendrequest model
-  Added validations and associations between models
- Add RSpec tests for each model
- Add integration tests for the site

### CODE REVIEW

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Impressive work @team :clap: :clap: However, some requirements need to be implemented in order to get approval for this project. Check :arrow_down: 

* [x] User can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] There is a button to invite user to friendship on users list page and on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] Add tests for the features above :arrow_up:  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by OdongAlican** [comment]
Hey @stratospherique. Thanks a lot for the feedback. We have implemented the requested changes. Please review it once again and give us your feedback

**Review by Forison — CHANGES_REQUESTED** [review]
## Status:: Change Required❌ 
# 🙋‍♂️ Team!!
#### Good job on addressing the issues raised earlier in the previous review, However, there is an issue we have to address with testing.
* [x] Kindly add test for all the public methods in the models ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/heroku.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- You are always encouraged to put the logic in the model to enable efficient testing. Currently, your unit test seems to only cover validations and associations but is equally important to test the public methods in the models.

**[OPTIONAL]**

* [x] It is better not to have the current user in the list of the users.👨
* [x] Kindly take a look at this [Article on ```N+1``` query in your spare time](https://medium.com/@bretdoucette/n-1-queries-and-how-to-avoid-them-a12f02345be5)
### That is it kindly make the changes and submit the project for a re-review🙇‍♂️.
#### Happy coding👨‍💻 👩‍💻
#### Addo Forison

**Comment by phylischumba** [comment]
Hello @Forison, thanks for the review. We have made the recommended changes. Kindly review it again and give us feedback

File: app/models/user.rb [hunk 1]
```diff
@@ -6,7 +6,60 @@ class User < ApplicationRecord
 
   validates :name, presence: true, length: { maximum: 20 }
 
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i.freeze
```
Comment (mosaaleb) [inline app/models/user.rb:9]: - [x] Devise already handles validating emails. Check `config/initializers/devise.rb` file, there is `config.email_regexp = /\A[^@\s]+@[^@\s]+\z/` option.

**Review by mosaaleb — APPROVED** [review]
### Excellent work @Phylis05 @OdongAlican :clap:

All requirements are handled. Tests cover all cases. 
Good job handling the requirements in previous reviews.

You can merge :+1:

##### Muhammad Ebeid

File: app/models/user.rb (hunk 1 above)
Comment (phylischumba) [reply app/models/user.rb:9]: Hey @mosaaleb, thanks for the review and the great feedback. We have implemented the change you requested.

**Pull request merged**